src/vsphere_cpi/lib/cloud/vsphere/nsxt_provider.rb

```diff
@@ -125,11 +125,16 @@ def add_vm_to_nsgroups(vm, ns_groups)
       lports = logical_ports(vm)
       nsgroups.each do |nsgroup|
         logger.info("Adding LogicalPorts: #{lports.map(&:id)} to NSGroup '#{nsgroup.id}'")
-        grouping_obj_svc.add_or_remove_ns_group_expression(
-          nsgroup.id,
-          *to_simple_expressions(lports),
-          'ADD_MEMBERS'
-        )
+        begin
+          grouping_obj_svc.add_or_remove_ns_group_expression(
+            nsgroup.id,
+            *to_simple_expressions(lports),
+            'ADD_MEMBERS'
+          )
+        rescue NSXT::ApiCallError => e
+          retry if e.code == 409 || e.code == 412 #Conflict or PreconditionFailed
+          raise e
+        end
```

Comment:
My only concern here is that there is no wait time between retries. This can severely overwhelm NSXT manager if lot of these calls are in play.
I know we decided initially to use simple retry, but can we implement this using bosh retryable and wait times between retries. like we use in function logical_ports?

Reply:
Would adding a `sleep(certain duration)` after rescuing the error and before retrying work?
Does using BOSH retryable give us some advantage over this?